Add core governance document (#7)
neoforged/governance: core.md

# Governance of the NeoForged project

> The NeoForged project is split between [github](https://github.com/NeoForged) and [discord](https://discord.gg/UvedJ9m). It is a community of volunteers, maintaining a free and open source ecosystem of modifications to the game Minecraft.

Reviewer: ```suggestion
The NeoForged project is split between [GitHub](https://github.com/NeoForged) and [Discord](https://discord.neoforged.net). It is a community of volunteers, maintaining a free and open source ecosystem of modifications to the Minecraft game.
```

### What this document is

This document is:
 * A non-exhaustive list of the working principles of the project.

>  * Intended to help direct the project in times of uncertainty.

Reviewer: Direct doesn't seem like a good fit for the sentence. Maybe steer?

Author: It's about project direction, so the verb "to direct" seems the best fit here.

This document is built on the fundamental principle of "people generally want to help, so we should not get in the way of it".

The core ethic is "people over processes", so this document is intentionally light on processes - situations that come up are rarely one-size-fits-all, and should be treated on a case by case basis, with the best process to fit the person.

### What this document is not

This document is not:
 * Intended to enforce rules on all members.
 * Intended to catch all possibilities.

>  * Intended to prevent "rules lawyering".

Reviewer: Are you sure it's *not* meant to *prevent* this phenomenon?

Author: Yes. The point of the document is to give guidance, not to have such excruciating detail that it'll prevent people trying to rules lawyer to governance.

It is not intended to prevent rules lawyering.

> The code of conduct of the project is taken to be the content of [the Contributor Covenant](https://www.contributor-covenant.org/version/2/1/code_of_conduct/), so this document is not the code of conduct.

Reviewer: ```suggestion
The Code of Conduct of the project is taken to be the content of [the Contributor Covenant](https://www.contributor-covenant.org/version/2/1/code_of_conduct/), so this document is not the Code of Conduct.
```
We should still probably have our own CoC

Author: If there's a pressing need over the Contributor Covenant, I can see that, yeah.

I'm not sure if there's anything we need that isn't covered, to be honest. Nobody is above the rules.

Conflict resolution is the responsibility of the Moderation and Steering Council teams. They may wish to append a section to this document, but currently, this document is not conflict resolution.

## Definitions

Following are some core definitions that will be referred to in the later sections of the document.

### The Project

> "The Project" refers to The NeoForged project, with a presence on both GitHub and Discord, as noted at the start of this document. It does not refer to any specific Repository or subproject.

Reviewer: If you capitalize Repository then you should do so for Subproject as well.

### Subprojects

Subprojects are large tasks that the Project itself is undergoing.

These may be:

* Repositories under the Project's GitHub
* Large, multi-repository tasks that are being undertaken (ie. updating the Modloader to a new version of the Game)

> * Smaller, more targetted tasks that are being undertaken (ie. refactoring the Modloader to remove or replace or rework a feature)

Reviewer: Considering the lack of punctuation these sentences should start with a lowercase letter and end in a semicolon.
Also, NeoForge is not a mod loader, it is a framework, and the framework needs to be updated, not FML.
Capitalize `repository` in `multi-repository` too.

Author: Right, yeah, i forgot that we split the loader out from the framework.

### Teams

The Project is split into four teams.

A team is a group of people with similar goals, and equal weighting in most decision making - with some exceptions, which will be explored later.

> You may [apply to join some teams](https://tinyurl.com/ForgeApplications), but some require appointment from within an existing team, or further voting after application.

Reviewer: The short URL should be renamed, and it should be under our domain (as is the Discord invite) so it can be updated later on if needs be. Additionally, which are those "some" teams (i.e. document the requirements in their specific subsections).

Author: Good call.

> #### Steering Council

Reviewer: This section should mention that the powers of the council are not all-encompassing and that decisions on different matters shall be first ran by the team responsible and may only be vetoed by the council under exceptional reasoned circumstances (“drama”, draws).

> The Steering Council are the management of the Project. 

Reviewer: Council is a singular noun.

Author: The people inside it are not, though; good catch nonetheless.

They "steer" the project to keep it aligned with the long-term vision, hence its secondary name of Vision Team.

The Council has at all times an odd number of active members, such that stalemates are not possible.

A member may choose to become inactive and appoint another in their place until they return.

> At least one member of the Council is promoted from other teams via an annual internal vote, to ensure that the Council always has a fresh opinion and does not stagnate over time.

Reviewer: The way this is worded would imply that the Council as a team has opinions. I would say it "benefits from fresh opinions"

Author: Fresh insight from a new team member, a different view on situations. Yes, good call.

Once a year, [an internal voting process](https://peps.python.org/pep-8104/) will determine the members that will be promoted to the Steering Council, via the ranked voting system. It is intended that at least one, and at most three members will be promoted at a time.

The Council also serves double duty as a mediator for disagreements between other teams. These shall be resolved by discussion, rather than by unanimous applications of the rules, so this is not in the purview of this document.

#### Maintainers

> The Maintainers team consists of all project developers who have push access or technical rights to the project's repositories or infrastructure.

Reviewer: Again, decide on repository or Repository.

They are the ones keeping the Project's repositories up to date, and ensuring the quality and standards of the Project are upheld, as well as adding new features as the community comes to require them.

> Though maintainers may individually or collectively have different access rights across the project's many repositories or infrastructure services, all maintainers within the team are considered equal to each other with respect to their rights within the team's governance.

Reviewer: ```suggestion
Though Maintainers may individually or collectively have different access rights across the Project's many Repositories or Infrastructure services, all Maintainers within the team are considered equal to each other with respect to their rights within the team's governance.
```

Maintainers may be appointed by other maintainers, or by application from the person who wishes to join the team.

#### Moderators

The Moderation team is composed of all moderators of all community spaces (GitHub, Discord) under the purview of the project.

They have the responsibility of enforcing the policies and rules of each space, as well as the Code of Conduct, and applying disciplinary sanctions to people who violate said rules.

#### Subproject Leads

A special group exists for those designated as "subproject leaders".

The members of this group were assigned from the Maintainers team, and is indeed a strict subset of the Maintainers team.

> The role of this team is to take lead of a subproject, as defined above.

Reviewer: "above" being? (link)

The subproject leader has total purview of their given subproject (repository or task) up to and including their primary goal.

They may:
* Decide the structure of their team, in whatever capacity is neccessary
* Decide who may or may not be on their team
* Delegate the power of subproject leader to another person
* Make final decisions on the implementation and structure of their subproject
* Overrule another Maintainer on the strict matter of their subproject, and no other.

They are not:
* A project dictator
* Unaccountable
* Free from scrutiny

They may not:
* Make decisions not related to their subproject

> * Make major architectural decisions without the assent of the Steering Council

Reviewer: Full stop at the end of each sentence or lowercase and semicolons.

> * Make major architectural decisions without the assent of the Steering Council

Reviewer: Why the Council directly and not the team assigned to the project first?

Author: I specifically meant major architectural decisions about the overall Project / framework, outside of the exclusive jurisdiction of the subproject.

# Governance of the NeoForged project

The NeoForged project is split between [GitHub](https://github.com/NeoForged) and [Discord](https://discord.neoforged.net). It is a community of volunteers, maintaining a free and open source ecosystem of modifications to the Minecraft game.

### What this document is

This document is:
 * A non-exhaustive list of the working principles of the project.
 * Intended to help guide the project in times of uncertainty.

This document is built on the fundamental principle of "people generally want to help, so we should not get in the way of it".

The core ethic is "people over processes", so this document is intentionally light on processes - situations that come up are rarely one-size-fits-all, and should be treated on a case by case basis, with the best process to fit the person.

### What this document is not

This document is not:
 * Intended to enforce rules on all members.
 * Intended to catch all possibilities.
 * Intended to prevent "rules lawyering".

> The Code of Conduct of the project is taken to be the content of [the Contributor Covenant](https://www.contributor-covenant.org/version/2/1/code_of_conduct/), so this document is not the Code of Conduct.

Reviewer 2: Maybe indicate that we are free to extend or overrule the covenant if we ever deem it to be needed. People can get annoyed if we do not clarify that ahead of time.

> Conflict resolution is the responsibility of the Moderation and Steering Council teams. They may wish to append a section to this document, but currently, this document is not conflict resolution.

Reviewer 2: Conflict resolution is not the responsibility of the moderation team IMHO.
They are the community managers and moderators but have no direct moderation power within the organization, that should be the steering council only.

Author: Resolving conflicts between community members is what I was going for, here. That's the job of the Moderators. It's part of the definition of "moderating" a discussion.

Reviewer 2: Specify that it is "Conflict resolution with the community is the responsibility of the moderation team, while conflict resolution within the organization is the responsibility of the steering council". This prevents any misunderstandings about the relevant roles.

## Definitions

Following are some core definitions that will be referred to in the later sections of the document.

### The Project

"The Project" refers to The NeoForged project, with a presence on both GitHub and Discord, as noted at the start of this document. It does not refer to any specific repository or subproject.

### Subprojects

Subprojects are large tasks that the Project itself is undergoing.

These may be:

* Repositories under the Project's GitHub.
* Large, multi-repository tasks that are being undertaken (ie. updating the API to a new version of the Game).
* Smaller, more targetted tasks that are being undertaken (ie. refactoring the Modloader or API to remove or replace or rework a feature).

### Teams

The Project is split into four teams.

A team is a group of people with similar goals, and equal weighting in most decision making - with some exceptions, which will be explored later.

You may [apply to join some teams](https://links.neoforged.net/apply), but some require appointment from within an existing team, or further voting after application.

#### Steering Council

The Steering Council is the overall management and oversight of the Project. 

> They "steer" the project to keep it aligned with the long-term vision, hence its secondary name of Vision Team.

Reviewer: Use smart quotes (`“”`)

Author: I have absolutely no idea how.

Reviewer: ```suggestion
They “steer” the project to keep it aligned with the long-term vision, hence its secondary name of Vision Team.
```

However, the Council does not have all-encompassing powers. They are the avenue of last resort, and the vast majority of decisions should be made by the team in charge, not by the Steering Council.

The Council has these powers to resolve the 1% of exceptional situations, not the 99%.

#### Maintainers

The Maintainers team consists of all project developers who have push access or technical rights to the project's repositories or infrastructure.

They are the ones keeping the Project's repositories up to date, and ensuring the quality and standards of the Project are upheld, as well as adding new features as the community comes to require them.

Though Maintainers may individually or collectively have different access rights across the project's many repositories or infrastructure services, all Maintainers within the team are considered equal to each other with respect to their rights within the team's governance.

Maintainers may be appointed by other Maintainers, or by application from the person who wishes to join the team.

> 

Reviewer 2: Possibly state that they have no community moderation powers, and that the moderators are supposed to manage the community and its expectations.

#### Moderators

The Moderation team is composed of all moderators of all community spaces (GitHub, Discord) under the purview of the project.

> They have the responsibility of enforcing the policies and rules of each space, as well as the Code of Conduct, and applying disciplinary sanctions to people who violate said rules.

Reviewer 2: Correct, but maybe stipulate that they do not influence technical decisions.

Reviewer: This will lead to the ”but what if a moderator is a maintainer?” question. I don't find such statement needed given the implications.

#### Subproject Leads

A special group exists for those designated as "subproject leaders".

The members of this group were assigned from the Maintainers team, and is indeed a strict subset of the Maintainers team.

The role of this team is to take lead of a subproject, [as defined above](#subprojects).

The subproject leader has total purview of their given subproject (repository or task) up to and including their primary goal.

They may:
* Decide the structure of their team, in whatever capacity is neccessary.
* Decide who may or may not be on their team.
* Delegate the power of subproject leader to another person.
* Make final decisions on the implementation and structure of their subproject.
* Overrule another Maintainer on the strict matter of their subproject, and no other.

They are not:
* A project dictator.
* Unaccountable.
* Free from scrutiny.

They may not:
* Make decisions not related to their subproject.
* Make major architectural decisions relating to the entire Project without the assent of the Steering Council as well as the Maintainers.

> Finally, the decisions made by the subproject leader and their team are, under all circumstances, subject to regular scrutiny and final review by the Maintainers and the Steering Council, to ensure the subproject still holds to the Project's values and quality standards.

Reviewer 2: IMHO: We should add something like: `However, continuously putting them under question is not allowed. They should be able to do their job and should regularly check in with the rest of the maintainers to collect feedback.`

The core ethic is "people over processes", so this document is intentionally light on processes - situations that come up are rarely one-size-fits-all, and should be treated on a case by case basis, with the best process to fit the person.

### What this document is not

This document is not:
 * Intended to enforce rules on all members.
 * Intended to catch all possibilities.
 * Intended to prevent "rules lawyering".

The Code of Conduct of the project is taken to be the content of [the Contributor Covenant](https://www.contributor-covenant.org/version/2/1/code_of_conduct/), so this document is not the Code of Conduct. However, this may be altered with a copy held and maintained by the Project, only with overwhelming majority support by the entire team.

Conflict resolution is the responsibility of the Moderation and Steering Council teams. They may wish to append a section to this document, but currently, this document is not conflict resolution.

## Definitions

Following are some core definitions that will be referred to in the later sections of the document.

### The Project

> "The Project" refers to The NeoForged project, with a presence on both GitHub and Discord, as noted at the start of this document. It does not refer to any specific repository or subproject.

Reviewer: ```suggestion
"The Project" refers to The NeoForged project, the organisation, with a presence on both GitHub and Discord, as noted at the start of this document. It does not refer to any specific repository or subproject.
```

They are the ones keeping the Project's repositories up to date, and ensuring the quality and standards of the Project are upheld, as well as adding new features as the community comes to require them.

Though Maintainers may individually or collectively have different access rights across the project's many repositories or infrastructure services, all Maintainers within the team are considered equal to each other with respect to their rights within the team's governance.

Maintainers may be appointed by other Maintainers, or by application from the person who wishes to join the team.

However, they have no jurisdiction or weight in moderating or managing the community, unless the Maintainer in question is also a Moderator.

#### Moderators

The Moderation team is composed of all moderators of all community spaces (GitHub, Discord) under the purview of the project.

They have the responsibility of enforcing the policies and rules of each space, as well as the Code of Conduct, and applying disciplinary sanctions to people who violate said rules.

> 

Reviewer 3: How are moderators appointed? Do they also have the same process as maintainers? Or are they slightly different?

Author: Good point! They share the same application system.

Conflict resolution is a duty split between the Moderators and Steering Council; the former for interpersonal disputes in and with the community, and the latter for development-related disputes in the Project itself. Either team may wish to append a section to this document, but currently, this document is not conflict resolution.

## Definitions

Following are some core definitions that will be referred to in the later sections of the document.

### The Project

"The Project" refers to The NeoForged project, with a presence on both GitHub and Discord, as noted at the start of this document. It does not refer to any specific repository or subproject.

### Subprojects

Subprojects are large tasks that the Project itself is undergoing.

These may be:

* Repositories under the Project's GitHub.
* Large, multi-repository tasks that are being undertaken (ie. updating the API to a new version of the Game).
* Smaller, more targetted tasks that are being undertaken (ie. refactoring the Modloader or API to remove or replace or rework a feature).

> 

Reviewer: Consider clarifying that the maintainers and moderators team don't have to be odd as the council will break ties.

Reviewer 4: They should usually not be asked for a full vote anyway, so I don't think it matters.

A member may choose to become inactive and appoint another in their place until they return.

At least one member of the Council is promoted from other teams via an annual internal vote, to ensure that the Council always has fresh insight from a new team member with a different view on situations, and thus does not stagnate over time.

Once a year, [an internal voting process](https://peps.python.org/pep-8104/) will determine the members that will be promoted to the Steering Council, via the ranked voting system. It is intended that at least one, and at most three members will be promoted at a time.

The Council also serves double duty as a mediator for disagreements between other teams. These shall be resolved by discussion, rather than by unanimous applications of the rules, so this is not in the purview of this document.

However, the Council does not have all-encompassing powers. They are the avenue of last resort, and the vast majority of decisions should be made by the team in charge, not by the Steering Council.

The Council has these powers to resolve the 1% of exceptional situations, not the 99%.

#### Maintainers

> The Maintainers team consists of all project developers who have push access or technical rights to the project's repositories or infrastructure.

Reviewer: What is "technical rights"?

Author: This one was taken from @sciwhiz12's draft, you'd have to ask him.

Reviewer 5: Technical rights/access means the user has access to technical configuration or operations. It can take many forms, according to their context:
- For GitHub repositories, it chiefly refers to Write permission and above; having the permission to push commits to branches.
- For services, it refers to being able to reconfigure, enable, or disable service components.
- For server access, it refers to the access to login and access a console or configuration page for a server.

It is qualified as "technical" to distinguish it from other forms of rights/access, such as moderation rights/access (the ability to moderate users and their contents).

They are the ones keeping the Project's repositories up to date, and ensuring the quality and standards of the Project are upheld, as well as adding new features as the community comes to require them.

Though Maintainers may individually or collectively have different access rights across the project's many repositories or infrastructure services, all Maintainers within the team are considered equal to each other with respect to their rights within the team's governance.

Maintainers may be appointed by other Maintainers, or by application from the person who wishes to join the team.

However, they have no jurisdiction or weight in moderating or managing the community, unless the Maintainer in question is also a Moderator.

#### Moderators

The Moderation team is composed of all moderators of all community spaces (GitHub, Discord) under the purview of the project.

They have the responsibility of enforcing the policies and rules of each space, as well as the Code of Conduct, and applying disciplinary sanctions to people who violate said rules.

Moderators may be appointed by the Steering Council, or by application from the person who wishes to join the team.

> However, they have no jurisdiction or weight in technical discussions, unless the Moderator in question is also a Maintainer.

Reviewer: This wording sounds as if their opinion won't be heard and simply dismissed. Maybe clarify that they're free to comment and their opinion will be considered just as of any other person's.

All contributors and team members are here for fun - our goal is to provide joy to the community, and to ourselves. 

We are here because it is fun, so this document ensures that the Project doesn't become so bureaucratic or so much of a chore that the fun is removed for anybody.

### What this document is

This document is:
 * A non-exhaustive list of the working principles of the project.
 * Intended to help guide the project in times of uncertainty.

This document is built on the fundamental principle of "people generally want to help, so we should not get in the way of it".

The core ethic is "people over processes", so this document is intentionally light on processes - situations that come up are rarely one-size-fits-all, and should be treated on a case by case basis, with the best process to fit the person.

### What this document is not

This document is not:
 * Intended to enforce rules on all members.
 * Intended to catch all possibilities.
 * Intended to prevent "rules lawyering".

> The Code of Conduct of the project is taken to be the content of [the Contributor Covenant](https://www.contributor-covenant.org/version/2/1/code_of_conduct/), so this document is not the Code of Conduct. However, this may be altered with a copy held and maintained by the Project, only with overwhelming majority support by the entire team.

Reviewer 6: I am against the adoption of the contributor covenant due to its usage of politically contentious language and associated concerning FAQ, but can be on board with an altered version.

Reviewer 4: What are you referring to in particular?

Reviewer 7: I disagree. Contributors covenant is a widely accepted document for operating within public online spheres. It's only "political" controversy is from those who wish not to be bound by some of it's terms.

The Council also serves double duty as a mediator for disagreements between other teams. These shall be resolved by discussion, rather than by unanimous applications of the rules, so this is not in the purview of this document.

However, the Council does not have all-encompassing powers. They are the avenue of last resort, and the vast majority of decisions should be made by the team in charge, not by the Steering Council.

The Council has these powers to resolve the 1% of exceptional situations, not the 99%.

#### Maintainers

The Maintainers team consists of all project developers who have push access or technical rights (Write permissions to a Project's GitHub repository, or access to the Project's servers) to the project's repositories or infrastructure.

They are the ones keeping the Project's repositories up to date, and ensuring the quality and standards of the Project are upheld, as well as adding new features as the community comes to require them.

> Though Maintainers may individually or collectively have different access rights across the project's many repositories or infrastructure services, all Maintainers within the team are considered equal to each other with respect to their rights within the team's governance.

Reviewer 6: Seemingly contradictory language, assuming "their rights within the team's governance" refers to the definition above regarding it being those "who have push access or technical rights to the project's repositories or infrastructure."

Reviewer 7: I don't see any contradiction here? You're going to have to be super extremely explicit and outline precisely what you mean.

Reviewer 6: > Maintainers may individually or collectively have different access rights

and

> Maintainers within the team are considered equal to each other with respect to their rights

within the same sentence.

Author: Access rights (permission to push to different repositories) and rights in the team (whose opinion is weighted more) are taken to be separate subjects entirely.

Reviewer 5: You left out a key part of that second sentence (emphasis mine):

> _[...]_ all Maintainers within the team are considered equal to each other with respect to their rights **within the team's governance.**

That is, while Maintainers hold different levels of access to resources, repositories, or infrastruture, they have the same 'loudness of voice' and rights (e.g. voting rights) with each other in the Maintainers team.

Reviewer 6: Shouldn't those who specialise in subprojects (such as subproject leaders) be given additional considerations based on their knowledge of a given area, rather than treating a new maintainer and an experienced specialist equally?

Author: Yes. This is why there's a separate topic for Subproject Leaders explicitly outlining their authority and rights exceeding standard maintainers.

Reviewer 7: OK. this looks to be resolved to me. Paint wasn't understanding the context of access rights to repositories vs rights within the organization.
